### PR TITLE
[TT-XLA][PyTorch] Training Workload for Llama 3.2 1B  - LoRA

### PR DESCRIPTION
**Pull request by mmilosevicTT**
### Ticket
Closes https://github.com/tenstorrent/tt-blacksmith/issues/232

### Model training setting

- **Model:** `Llama 3.2 1B`
- **Strategy:** SFT with LoRA layers on a single chip (N150)
- **Framework:** PyTorch
- **Task:** Sentiment prediction using `SST2`
 
### What's changed
This PR touches on the following things:
- Refactored `llama` experiment to support both frontends (TT-FFE and TT-XLA)
- Implemented `llama` training workload through TT-XLA

### Important findings

- Using Hugging Face loss leads to hangs in first epoch after 87 batches of training for TT-XLA PyTorch implementation on TT device.
- Using custom loss leads to hangs in first epoch after 265 batches of training for TT-XLA PyTorch implementation on TT device. 
- Using Hugging Face model loss on TT device does not yield correct results. The problem is that Hugging Face model loss automatically slices inputs and labels for Casual LM tasks, which does not translate to TT device.
- Using Hugging Face `tokenizer` is not working properly with tensors on TT device. To resolve printing of data, tensors need to be sent to CPU before decoding. 
- Calling `loss.item()` after `loss.backward()` forces collection of gradients and that is why we had a problem of TT devices looking like they are training twice as fast.
- One more problem that appears is that environment setup for GPU is not working. Importing PJRT plugin makes a problem for CUDA runs. This needs to be investigated and potentially fixed in setup.

### Experiment results

For completeness I am providing a train/loss curve of the experiments run during testing.

W B Chart 24_10_2025, 14_26_59

- tt-llama-test-no-loss-item-problem - Experiment without `loss.item()` after `.backward()`
- tt-llama-test-optimizer-cpu - Experiment with optimizer on CPU, but `loss.item()` after `.backward()`
- tt-llama-test-optimizer-device - Experiment with optimizer on TT device, but `loss.item()` after `.backward()`
- tt-llama-hf-loss-test - Experiment with hugging face loss
- gpu-llama-hf-loss-test - Experiment with hugging face loss on GPU
- gpu-llama-test - Experiment with custom loss on GPU
- cpu-llama-hf-loss-test - Experiment with hugging face loss on CPU
- cpu-llama-test - Experiment with custom loss on CPU

### Checklist
- [x] Main training workflow
- [x] W&B logging support
- [x] CPU testing
- [x] GPU testing
- [x] TT device testing

### CODE REVIEW

File: blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+from blacksmith.datasets.torch.llama.sst_dataset import SSTDataset
+from blacksmith.experiments.torch.llama.configs import TrainingConfig
+from blacksmith.models.torch.huggingface.hf_models import get_model, TextModelWrapper
+from blacksmith.tools.cli import generate_config
+
+
+def show_examples(examples, tokenizer):
+
+    for i, example in enumerate(examples):
+        if i > 10:
```
Comment (vladimirjovanovicTT) [inline blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py]: this seems a bit random :)

File: blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    print(f"\n=== Starting Validation ===")
+    model.eval()
+    total_val_loss = 0.0
+    num_val_batches = 0
+    collected_examples = []
+    max_examples = 10
+
+    with torch.no_grad():
+        for batch in tqdm(val_data_loader, desc="Validation"):
+            if num_val_batches > 10:
```
Comment (vladimirjovanovicTT) [inline blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py]: similar situation as in show_examples.

File: blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            expected_output = batch["labels"].to(device)
+
+            # Forward pass
+            outputs = model(input_ids=input_ids, attention_mask=attention_mask)
+            logits = outputs.logits
+
+            # Shift logits and labels for causal LM: predict next token
+            # logits[:, :-1] predicts tokens at positions 1:, so compare with labels[:, 1:]
+            shift_logits = logits[:, :-1, :].contiguous()
+            shift_labels = expected_output[:, 1:].contiguous()
```
Comment (vladimirjovanovicTT) [inline blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py]: It would be cleaner if the output labels are shifted in the data loader.

File: blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    running_loss = 0.0
+
+    try:
+        for epoch in range(config.num_epochs):
+            print(f"\n=== Epoch {epoch + 1}/{config.num_epochs} ===")
+
+            for batch in tqdm(train_data_loader, desc="Training"):
+                # Set model to train mode
+                # We need to do this in inner loop to avoid validation on step logging from switching to eval mode
+                model.train()
```
Comment (vladimirjovanovicTT) [inline blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py]: nit: we do not need to do this on every step, since it is reset only in validation.

File: blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+                    if config.save_strategy == "steps":
+                        checkpoint_path = os.path.join(
+                            config.output_dir, "checkpoints", f"checkpoint-{global_step}.pth"
+                        )
+                        torch.save(model.state_dict(), checkpoint_path)
+
+            if config.save_strategy == "epoch":
+                checkpoint_path = os.path.join(config.output_dir, "checkpoints", f"checkpoint-{epoch+1}.pth")
+                torch.save(model.state_dict(), checkpoint_path)
+
```
Comment (vladimirjovanovicTT) [inline blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py]: We are missing early stopping mechanism and fallback to best validation loss chkpt. I'm good with having the PR as is and addressing this as part of experiment consolidation effort.

File: blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py (hunk 4 above)
Comment (mmilosevicTT) [reply blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py]: Moved to switch model only when we do validation and at the start of training for now.

File: blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py (hunk 3 above)
Comment (mmilosevicTT) [reply blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py]: Moved to collate_fn to shift labels. For logits we need to keep it like this.

File: blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    for i, example in enumerate(examples):
+        if i > 10:
+            break
+        print(f"\nExample {i+1} (from batch {example['batch_num']}):")
+
+        input_ids = example["input_ids"].to("cpu")
+        expected = example["expected"].to("cpu")
+        predicted = example["predicted"].to("cpu")
+
+        valid_mask = expected != -100
```
[the remark this answers is not in the document]
Comment (mmilosevicTT) [reply blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py]: Well we want to decode the value, no? How do you interpret `-100` in decode method when it represents all ignored states? This check is just to keep the code safe from crashing on printing examples.

File: blacksmith/experiments/torch/llama/configs.py [hunk 7]
```diff
@@ -43,4 +43,5 @@ class TrainingConfig(BaseModel):
     save_total_limit: int = Field(default=3, gt=0)
     do_train: bool = Field(default=True)
```
[the remark this answers is not in the document]
Comment (mmilosevicTT) [reply blacksmith/experiments/torch/llama/configs.py]: Removed unused configs

**Review by pglusacTT — APPROVED** [review]

File: blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+
+    shifted_labels = labels[:, 1:].contiguous()
+
+    return {"input_ids": input_ids, "attention_mask": attention_mask, "labels": shifted_labels}
+
+
+def train(config, device):
+    # Get model
+    model = get_model(config)
+    model = model.to(device)
```
[the remark this answers is not in the document]
Comment (mmilosevicTT) [reply blacksmith/experiments/torch/llama/xla/test_llama_fine_tuning_pure_torch.py]: Removed.
